Added extra check in reshape for shapes whose less is less than 2 (#15200 fix)

ajakovljevicTT opened the pull request:
[Link to Github Issue](https://github.com/tenstorrent/tt-metal/issues/15200)

### Problem description:
Due to a tiling check in ReshapeViewOperation::invoke which indexes -2 index of the shape of the tensor, the check fails if the shape rank is 1, this MR just adds an additional check.

### What's changed
Added an additional check in `ReshapeViewOperation::invoke`

ajakovljevicTT committed:
Added extra check in reshape

ntarafdar commented on ttnn/cpp/ttnn/operations/data_movement/reshape_view/reshape.cpp:
```diff
@@ -111,7 +111,7 @@ ttnn::Tensor ReshapeViewOperation::invoke(const ttnn::Tensor& tensor, const ttnn
+        ((shape.with_tile_padding().rank() < 2 or shape.with_tile_padding()[-2] % ttnn::TILE_SIZE == 0) and (shape.with_tile_padding()[-1] % ttnn::TILE_SIZE == 0)) and
         (tensor_shape.with_tile_padding()[-1] == shape.with_tile_padding()[-1])
         );
 
```
@jvegaTT could you look at this?

jvegaTT replied:
I don't think this is valid. Say I reshape {2,1[32],16[32]} to [32] which is a valid reshape. This goes from needing 2 tiles to needing 1 so a view is not legal. But 32 in tile is really {1,1[32],32[32]} and the above logic would have said that a view is possible. Then the tensor.reshape will assert a failure.

jvegaTT replied:
On update - that's better but I will need to do a better fix for the second case later. I will open up a ne issue for that to remember

jvegaTT replied:
Tiny readability thing, you now have:
cond a and
(cond b and cond c) and
cond d

I think it will be easier to follow as
cond a and
cond b and 
cond c and
cond d

I know the old code did this but just a thought though not necessary

ajakovljevicTT committed:
Quick fix

jvegaTT reviewed (changes requested):
Wait sorry I just realized that this will also cause an issue in the new this_is_view check just before. To avoid needing a second PR could you change the this_is_view equation?  I will write in a reply the new equation

jvegaTT commented:
> Wait sorry I just realized that this will also cause an issue in the new this_is_view check just before. To avoid needing a second PR could you change the this_is_view equation? I will write in a reply the new equation

![image](https://github.com/user-attachments/assets/9ab8970d-4718-48ff-b5a4-92e9165f916b)

jvegaTT commented:
> Wait sorry I just realized that this will also cause an issue in the new this_is_view check just before. To avoid needing a second PR could you change the this_is_view equation? I will write in a reply the new equation



> Wait sorry I just realized that this will also cause an issue in the new this_is_view check just before. To avoid needing a second PR could you change the this_is_view equation? I will write in a reply the new equation

First please merge in my changes when I added the view op, then modify my this_is_view check as follows.

![image](https://github.com/user-attachments/assets/6f3d3a6a-b89e-4164-afe4-839068980509)

ajakovljevicTT committed:
Rebase changes

ajakovljevicTT committed:
Added rank check

ajakovljevicTT commented:
> > Wait sorry I just realized that this will also cause an issue in the new this_is_view check just before. To avoid needing a second PR could you change the this_is_view equation? I will write in a reply the new equation
> 
> > Wait sorry I just realized that this will also cause an issue in the new this_is_view check just before. To avoid needing a second PR could you change the this_is_view equation? I will write in a reply the new equation
> 
> First please merge in my changes when I added the view op, then modify my this_is_view check as follows (with colons rather than commas, sorry for the bad syntax
> 
> ![image](https://private-user-images.githubusercontent.com/183505079/388622078-6f3d3a6a-b89e-4164-afe4-839068980509.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzIyNjUxNjQsIm5iZiI6MTczMjI2NDg2NCwicGF0aCI6Ii8xODM1MDUwNzkvMzg4NjIyMDc4LTZmM2QzYTZhLWI4OWUtNDE2NC1hZmU0LTgzOTA2ODk4MDUwOS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTIyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyMlQwODQxMDRaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTZlZTA5MjJhODdhZGQwNjZlYzFhZGZiYTA1YTZkNDE2ZGEyYjUxZWM5MTEyYjAxMGI3MWExZjIyOGQyZjdkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.1sXcOPLfkKoeZm0kiVbusJY1y2Dg3ZFWbfVj5dB4x28)

@jvegaTT I have added the checks as you have wrote. However, I still got the error when trying to reshape 1-dimensional tensors, as it falls down to the `PerformView` path (the error is: `Alignment should have at least 2 dimensions for Tile layout`). So I added an additional check to let it falls to the catch-all path `convert_tensor_to_rm_reshape_convert_back_to_orig_layout` and tested if it works. Let me know if this is ok, thanks!

jvegaTT commented:
That's not good. To give you context the OP is about 1000 times faster when this condition is taken. Can you investigate a little bit to see if there is an easy fix? If not we can let this PR through but please raise a github issue and assign me to it to resolve it.

ajakovljevicTT committed:
Changed shapes

jvegaTT reviewed (approved):
Thanks for all the changes. It looks good!
